src/Analyzer.Utilities/JsonLineNumberResolver.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Text.RegularExpressions;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Templates.Analyzer.Utilities
+{
+    /// <summary>
+    /// An <c>ILineNumberResolver</c> used for resolving line numbers from an expanded JSON template to the original JSON template.
+    /// </summary>
+    public class JsonLineNumberResolver : IJsonLineNumberResolver
+    {
+        private static readonly Regex resourceIndexInPath = new Regex(@"Resources\[(?<index>\d+)\]", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+
+        /// <summary>
+        /// Given a JSON path in an expanded JSON template, find the equivalent line number
+        /// in the original JSON template.
+        /// </summary>
+        /// <param name="pathInExpandedTemplate">The path in the expanded template
+        /// to find the line number of in the original template.</param>
+        /// <param name="expandedTemplateRoot">The root of the expanded template.</param>
+        /// <param name="originalTemplateRoot">The root of the original template.</param>
+        /// <returns>The line number of the equivalent location in the original template,
+        /// or 0 if it can't be determined.</returns>
+        public int ResolveLineNumberForOriginalTemplate(
+            string pathInExpandedTemplate,
+            JToken expandedTemplateRoot,
+            JToken originalTemplateRoot)
+        {
+            /*
```

Comment:
Would moving this explanation to a .md file be better?  It's quite long.  We don't really have a "dev docs" section anywhere yet.  I could just add the file to the Analyzer.Utilities directory, or we could start a dev docs location somewhere.

Reply:
I ended up removing this block entirely.  After refactoring, I think the code is clear enough to no longer need this long explanation.